Telemetry frequently lost during application start-up when sidecar OTEL collector isn't ready yet, even with retry feature enabled

**Describe the bug**
We're running java applications in kubernetes with an OpenTelemetry Collector sidecar container that an application container exports to. During pod start-up we're occasionally seeing failed exports (based on `otlp_exporter_exported{success="false"}`). We do have `OTEL_EXPERIMENTAL_EXPORTER_OTLP_RETRY_ENABLED` enabled.

I plotted some of the SDK metrics for a pod where this happened:

<img width="1501" alt="image" src="https://github.com/user-attachments/assets/249c9e56-3c93-4a70-ae81-5525db4c5a43">

It doesn't look like the errors are retried based on the error logs and a brief look at the gRPC exporter code? Also tested this with the HTTP exporter as well.

```
[otel.javaagent 2024-07-31 21:19:36:628 +0000] [OkHttp http://localhost:4318/...] ERROR io.opentelemetry.exporter.internal.http.HttpExporter - Failed to export spans. The request could not be executed. Full error message: Failed to connect to localhost/[0:0:0:0:0:0:0:1]:4318 
java.net.ConnectException: Failed to connect to localhost/[0:0:0:0:0:0:0:1]:4318
	at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.kt:297)
	at okhttp3.internal.connection.RealConnection.connect(RealConnection.kt:207)
	at okhttp3.internal.connection.ExchangeFinder.findConnection(ExchangeFinder.kt:226)
	at okhttp3.internal.connection.ExchangeFinder.findHealthyConnection(ExchangeFinder.kt:106)
	at okhttp3.internal.connection.ExchangeFinder.find(ExchangeFinder.kt:74)
	at okhttp3.internal.connection.RealCall.initExchange$okhttp(RealCall.kt:255)
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.kt:32)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.kt:95)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.kt:83)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.kt:76)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at io.opentelemetry.exporter.sender.okhttp.internal.RetryInterceptor.intercept(RetryInterceptor.java:83)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:201)
	at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:517)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
	Suppressed: java.net.ConnectException: Failed to connect to localhost/127.0.0.1:4318
		... 21 more
	Caused by: java.net.ConnectException: Connection refused
		at java.base/sun.nio.ch.Net.pollConnect(Native Method)
		at java.base/sun.nio.ch.Net.pollConnectNow(Net.java:672)
		at java.base/sun.nio.ch.NioSocketImpl.timedFinishConnect(NioSocketImpl.java:542)
		at java.base/sun.nio.ch.NioSocketImpl.connect(NioSocketImpl.java:597)
		at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:327)
		at java.base/java.net.Socket.connect(Socket.java:633)
		at okhttp3.internal.platform.Platform.connectSocket(Platform.kt:128)
		at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.kt:295)
		... 20 more
Caused by: java.net.ConnectException: Connection refused
	at java.base/sun.nio.ch.Net.pollConnect(Native Method)
	at java.base/sun.nio.ch.Net.pollConnectNow(Net.java:672)
	at java.base/sun.nio.ch.NioSocketImpl.timedFinishConnect(NioSocketImpl.java:542)
	at java.base/sun.nio.ch.NioSocketImpl.connect(NioSocketImpl.java:597)
	at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:327)
	at java.base/java.net.Socket.connect(Socket.java:633)
	at okhttp3.internal.platform.Platform.connectSocket(Platform.kt:128)
	at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.kt:295)
	... 20 more
```

```
[otel.javaagent] [OkHttp http://localhost:4317/...] WARN io.opentelemetry.exporter.internal.grpc.GrpcExporter - Failed to export metrics. Server responded with gRPC status code 2. Error message: Failed to connect to localhost/[0:0:0:0:0:0:0:1]:4317

(forgot to capture stacktrace)
```
My guess is there's some sort of race condition where the sidecar collector isn't ready for the export. The `queueSize` isn't reported as hitting the max that triggers an export. It's really close to that maximum though, so maybe a gauge would miss recording when it hits that threshold?

**Steps to reproduce**
Probably can reproduce this by exporting without a collector, since I think it'll hit the same error

**What did you expect to see?**
It'd be nice if the SDK retired the request since the sidecar comes up pretty quickly. 

Maybe the SDK retries on these errors for up to X seconds after application start-up. Always retrying on these error might backfire, like during pod termination when the OTEL collector is terminated before the application container

**What did you see instead?**
Failed exports

**What version and what artifacts are you using?**
We're using version 2.0.0 of the OpenTelemetry java agent

**Environment**
      - name: OTEL_EXPORTER_OTLP_PROTOCOL
        value: http/protobuf
      - name: OTEL_EXPORTER_OTLP_METRICS_TEMPORALITY_PREFERENCE
        value: cumulative
      - name: OTEL_EXPERIMENTAL_METRICS_CARDINALITY_LIMIT
        value: 10000
      - name: OTEL_SPAN_ATTRIBUTE_VALUE_LENGTH_LIMIT
        value: "3000"
      - name: OTEL_SPAN_ATTRIBUTE_COUNT_LIMIT
        value: "60"
      - name: OTEL_BSP_MAX_QUEUE_SIZE
        value: "10000"
      - name: OTEL_METRIC_EXPORT_INTERVAL
        value: "15000"
      - name: OTEL_METRIC_EXPORT_TIMEOUT
        value: "10000"
      - name: OTEL_METRICS_EXPORTER
        value: none
      - name: OTEL_LOGS_EXPORTER
        value: none
      - name: OTEL_EXPERIMENTAL_EXPORTER_OTLP_RETRY_ENABLED
        value: "true"

--- a/exporters/sender/jdk/src/main/java/io/opentelemetry/exporter/sender/jdk/internal/JdkHttpSender.java
+++ b/exporters/sender/jdk/src/main/java/io/opentelemetry/exporter/sender/jdk/internal/JdkHttpSender.java
@@ -5,6 +5,8 @@
 
 package io.opentelemetry.exporter.sender.jdk.internal;
 
+import static java.util.stream.Collectors.joining;
+
 import io.opentelemetry.exporter.internal.compression.Compressor;
 import io.opentelemetry.exporter.internal.http.HttpSender;
 import io.opentelemetry.exporter.internal.marshal.Marshaler;
@@ -25,6 +27,7 @@ import java.time.Duration;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.StringJoiner;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ConcurrentLinkedQueue;
 import java.util.concurrent.ExecutorService;
@@ -33,6 +36,8 @@ import java.util.concurrent.ThreadLocalRandom;
 import java.util.concurrent.TimeUnit;
 import java.util.function.Consumer;
 import java.util.function.Supplier;
+import java.util.logging.Level;
+import java.util.logging.Logger;
 import javax.annotation.Nullable;
 import javax.net.ssl.SSLContext;
 import javax.net.ssl.SSLException;
@@ -51,6 +56,8 @@ public final class JdkHttpSender implements HttpSender {
       ThreadLocal.withInitial(NoCopyByteArrayOutputStream::new);
   private static final ThreadLocal<ByteBufferPool> threadLocalByteBufPool =
       ThreadLocal.withInitial(ByteBufferPool::new);
+
+  private static final Logger logger = Logger.getLogger(JdkHttpSender.class.getName());
 
   private final ExecutorService executorService = Executors.newFixedThreadPool(5);
   private final HttpClient client;
@@ -211,11 +218,37 @@ public final class JdkHttpSender implements HttpSender {
         exception = e;
       }
 
-      if (httpResponse != null && !retryableStatusCodes.contains(httpResponse.statusCode())) {
-        return httpResponse;
+      if (httpResponse != null) {
+        boolean retryable = retryableStatusCodes.contains(httpResponse.statusCode());
+        if (logger.isLoggable(Level.FINER)) {
+          logger.log(
+              Level.FINER,
+              "Attempt "
+                  + attempt
+                  + " returned "
+                  + (retryable ? "retryable" : "non-retryable")
+                  + " response: "
+                  + responseStringRepresentation(httpResponse));
+        }
+        if (!retryable) {
+          return httpResponse;
+        }
       }
-      if (exception != null && !isRetryableException(exception)) {
-        throw exception;
+      if (exception != null) {
+        boolean retryable = isRetryableException(exception);
+        if (logger.isLoggable(Level.FINER)) {
+          logger.log(
+              Level.FINER,
+              "Attempt "
+                  + attempt
+                  + " failed with "
+                  + (retryable ? "retryable" : "non-retryable")
+                  + " exception",
+              exception);
+        }
+        if (!retryable) {
+          throw exception;
+        }
       }
     } while (attempt < retryPolicy.getMaxAttempts());
 
@@ -223,6 +256,17 @@ public final class JdkHttpSender implements HttpSender {
       return httpResponse;
     }
     throw exception;
+  }
+
+  private static String responseStringRepresentation(HttpResponse<?> response) {
+    StringJoiner joiner = new StringJoiner(",", "HttpResponse{", "}");
+    joiner.add("code=" + response.statusCode());
+    joiner.add(
+        "headers="
+            + response.headers().map().entrySet().stream()
+                .map(entry -> entry.getKey() + "=" + String.join(",", entry.getValue()))
+                .collect(joining(",", "[", "]")));
+    return joiner.toString();
   }
 
   private void write(Marshaler marshaler, OutputStream os) throws IOException {

--- a/exporters/sender/jdk/src/test/java/io/opentelemetry/exporter/sender/jdk/internal/JdkHttpSenderTest.java
+++ b/exporters/sender/jdk/src/test/java/io/opentelemetry/exporter/sender/jdk/internal/JdkHttpSenderTest.java
@@ -82,7 +82,8 @@ class JdkHttpSenderTest {
         new JdkHttpSender(
             mockHttpClient,
             // Connecting to localhost on an unused port address to trigger
-            // java.net.ConnectException
+            // java.net.ConnectException (or java.net.http.HttpConnectTimeoutException on linux java
+            // 11+)
             "http://localhost:" + freePort(),
             null,
             false,
@@ -95,7 +96,12 @@ class JdkHttpSenderTest {
                 .build());
 
     assertThatThrownBy(() -> sender.sendInternal(new NoOpMarshaler()))
-        .isInstanceOf(ConnectException.class);
+        .satisfies(
+            e ->
+                assertThat(
+                        (e instanceof ConnectException)
+                            || (e instanceof HttpConnectTimeoutException))
+                    .isTrue());
 
     verify(mockHttpClient, times(2)).send(any(), any());
   }

--- a/exporters/sender/jdk/src/test/java/io/opentelemetry/exporter/sender/jdk/internal/JdkHttpSenderTest.java
+++ b/exporters/sender/jdk/src/test/java/io/opentelemetry/exporter/sender/jdk/internal/JdkHttpSenderTest.java
@@ -18,6 +18,8 @@ import io.opentelemetry.exporter.internal.marshal.Marshaler;
 import io.opentelemetry.exporter.internal.marshal.Serializer;
 import io.opentelemetry.sdk.common.export.RetryPolicy;
 import java.io.IOException;
+import java.net.ConnectException;
+import java.net.ServerSocket;
 import java.net.http.HttpClient;
 import java.net.http.HttpConnectTimeoutException;
 import java.time.Duration;
@@ -53,8 +55,8 @@ class JdkHttpSenderTest {
     sender =
         new JdkHttpSender(
             mockHttpClient,
-            "http://10.255.255.1", // Connecting to a non-routable IP address to trigger connection
-            // timeout
+            // Connecting to a non-routable IP address to trigger connection timeout
+            "http://10.255.255.1",
             null,
             false,
             "text/plain",
@@ -72,6 +74,38 @@ class JdkHttpSenderTest {
         .isInstanceOf(HttpConnectTimeoutException.class);
 
     verify(mockHttpClient, times(2)).send(any(), any());
+  }
+
+  @Test
+  void sendInternal_RetryableConnectException() throws IOException, InterruptedException {
+    sender =
+        new JdkHttpSender(
+            mockHttpClient,
+            // Connecting to localhost on an unused port address to trigger
+            // java.net.ConnectException
+            "http://localhost:" + freePort(),
+            null,
+            false,
+            "text/plain",
+            Duration.ofSeconds(10).toNanos(),
+            Collections::emptyMap,
+            RetryPolicy.builder()
+                .setMaxAttempts(2)
+                .setInitialBackoff(Duration.ofMillis(1))
+                .build());
+
+    assertThatThrownBy(() -> sender.sendInternal(new NoOpMarshaler()))
+        .isInstanceOf(ConnectException.class);
+
+    verify(mockHttpClient, times(2)).send(any(), any());
+  }
+
+  private static int freePort() {
+    try (ServerSocket socket = new ServerSocket(0)) {
+      return socket.getLocalPort();
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
   }
 
   @Test

--- a/exporters/sender/okhttp/src/main/java/io/opentelemetry/exporter/sender/okhttp/internal/RetryInterceptor.java
+++ b/exporters/sender/okhttp/src/main/java/io/opentelemetry/exporter/sender/okhttp/internal/RetryInterceptor.java
@@ -149,8 +149,7 @@ public final class RetryInterceptor implements Interceptor {
     if (e instanceof SocketTimeoutException) {
       String message = e.getMessage();
       // Connect timeouts can produce SocketTimeoutExceptions with no message, or with "connect
-      // timed
-      // out"
+      // timed out"
       return message == null || message.toLowerCase(Locale.ROOT).contains("connect timed out");
     } else if (e instanceof ConnectException) {
       // Exceptions resemble: java.net.ConnectException: Failed to connect to

--- a/exporters/sender/okhttp/src/main/java/io/opentelemetry/exporter/sender/okhttp/internal/RetryInterceptor.java
+++ b/exporters/sender/okhttp/src/main/java/io/opentelemetry/exporter/sender/okhttp/internal/RetryInterceptor.java
@@ -5,13 +5,19 @@
 
 package io.opentelemetry.exporter.sender.okhttp.internal;
 
+import static java.util.stream.Collectors.joining;
+
 import io.opentelemetry.sdk.common.export.RetryPolicy;
 import java.io.IOException;
+import java.net.ConnectException;
 import java.net.SocketTimeoutException;
 import java.util.Locale;
+import java.util.StringJoiner;
 import java.util.concurrent.ThreadLocalRandom;
 import java.util.concurrent.TimeUnit;
 import java.util.function.Function;
+import java.util.logging.Level;
+import java.util.logging.Logger;
 import okhttp3.Interceptor;
 import okhttp3.Response;
 
@@ -22,6 +28,8 @@ import okhttp3.Response;
  * at any time.
  */
 public final class RetryInterceptor implements Interceptor {
+
+  private static final Logger logger = Logger.getLogger(RetryInterceptor.class.getName());
 
   private final RetryPolicy retryPolicy;
   private final Function<Response, Boolean> isRetryable;
@@ -84,12 +92,39 @@ public final class RetryInterceptor implements Interceptor {
       } catch (IOException e) {
         exception = e;
       }
-      if (response != null && !Boolean.TRUE.equals(isRetryable.apply(response))) {
-        return response;
+      if (response != null) {
+        boolean retryable = Boolean.TRUE.equals(isRetryable.apply(response));
+        if (logger.isLoggable(Level.FINER)) {
+          logger.log(
+              Level.FINER,
+              "Attempt "
+                  + attempt
+                  + " returned "
+                  + (retryable ? "retryable" : "non-retryable")
+                  + " response: "
+                  + responseStringRepresentation(response));
+        }
+        if (!retryable) {
+          return response;
+        }
       }
-      if (exception != null && !Boolean.TRUE.equals(isRetryableException.apply(exception))) {
-        throw exception;
+      if (exception != null) {
+        boolean retryable = Boolean.TRUE.equals(isRetryableException.apply(exception));
+        if (logger.isLoggable(Level.FINER)) {
+          logger.log(
+              Level.FINER,
+              "Attempt "
+                  + attempt
+                  + " failed with "
+                  + (retryable ? "retryable" : "non-retryable")
+                  + " exception",
+              exception);
+        }
+        if (!retryable) {
+          throw exception;
+        }
       }
+
     } while (attempt < retryPolicy.getMaxAttempts());
 
     if (response != null) {
@@ -98,15 +133,31 @@ public final class RetryInterceptor implements Interceptor {
     throw exception;
   }
 
+  private static String responseStringRepresentation(Response response) {
+    StringJoiner joiner = new StringJoiner(",", "Response{", "}");
+    joiner.add("code=" + response.code());
+    joiner.add(
+        "headers="
+            + response.headers().toMultimap().entrySet().stream()
+                .map(entry -> entry.getKey() + "=" + String.join(",", entry.getValue()))
+                .collect(joining(",", "[", "]")));
+    return joiner.toString();
+  }
+
   // Visible for testing
   static boolean isRetryableException(IOException e) {
-    if (!(e instanceof SocketTimeoutException)) {
-      return false;
+    if (e instanceof SocketTimeoutException) {
+      String message = e.getMessage();
+      // Connect timeouts can produce SocketTimeoutExceptions with no message, or with "connect
+      // timed
+      // out"
+      return message == null || message.toLowerCase(Locale.ROOT).contains("connect timed out");
+    } else if (e instanceof ConnectException) {
+      // Exceptions resemble: java.net.ConnectException: Failed to connect to
+      // localhost/[0:0:0:0:0:0:0:1]:62611
+      return true;
     }
-    String message = e.getMessage();
-    // Connect timeouts can produce SocketTimeoutExceptions with no message, or with "connect timed
-    // out"
-    return message == null || message.toLowerCase(Locale.ROOT).contains("connect timed out");
+    return false;
   }
 
   // Visible for testing

--- a/exporters/sender/okhttp/src/test/java/io/opentelemetry/exporter/sender/okhttp/internal/RetryInterceptorTest.java
+++ b/exporters/sender/okhttp/src/test/java/io/opentelemetry/exporter/sender/okhttp/internal/RetryInterceptorTest.java
@@ -23,6 +23,8 @@ import com.linecorp.armeria.common.HttpStatus;
 import com.linecorp.armeria.testing.junit5.server.mock.MockWebServerExtension;
 import io.opentelemetry.sdk.common.export.RetryPolicy;
 import java.io.IOException;
+import java.net.ConnectException;
+import java.net.ServerSocket;
 import java.net.SocketTimeoutException;
 import java.time.Duration;
 import java.util.concurrent.TimeUnit;
@@ -155,6 +157,34 @@ class RetryInterceptorTest {
     verify(isRetryableException, times(5)).apply(any());
     // Should retry maxAttempts, and sleep maxAttempts - 1 times
     verify(sleeper, times(4)).sleep(anyLong());
+  }
+
+  @Test
+  void connectException() throws Exception {
+    client = connectTimeoutClient();
+    when(random.get(anyLong())).thenReturn(1L);
+    doNothing().when(sleeper).sleep(anyLong());
+
+    // Connecting to localhost on an unused port address to trigger java.net.ConnectException
+    int openPort = freePort();
+    assertThatThrownBy(
+            () ->
+                client
+                    .newCall(new Request.Builder().url("http://localhost:" + openPort).build())
+                    .execute())
+        .isInstanceOf(ConnectException.class);
+
+    verify(isRetryableException, times(5)).apply(any());
+    // Should retry maxAttempts, and sleep maxAttempts - 1 times
+    verify(sleeper, times(4)).sleep(anyLong());
+  }
+
+  private static int freePort() {
+    try (ServerSocket socket = new ServerSocket(0)) {
+      return socket.getLocalPort();
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
   }
 
   @Test

